src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -2169,198 +2158,117 @@ public BigDecimal sqrt(MathContext mc) {
-                    targetPrecision = originalPrecision;
-                    break;
+                    result = result.adjustToPreferredScale(preferredScale, mc.precision);
+                    return result;
                 }
-            }
 
-            // When setting the precision to use inside the Newton
-            // iteration loop, take care to avoid the case where the
-            // precision of the input exceeds the requested precision
-            // and rounding the input value too soon.
-            BigDecimal approx = guess;
-            int workingPrecision = working.precision();
-            do {
-                int tmpPrecision = Math.max(Math.max(guessPrecision, targetPrecision + 2),
-                                           workingPrecision);
-                MathContext mcTmp = new MathContext(tmpPrecision, RoundingMode.HALF_EVEN);
-                // approx = 0.5 * (approx + fraction / approx)
-                approx = ONE_HALF.multiply(approx.add(working.divide(approx, mcTmp), mcTmp));
-                guessPrecision *= 2;
-            } while (guessPrecision < targetPrecision + 2);
-
-            BigDecimal result;
-            RoundingMode targetRm = mc.getRoundingMode();
-            if (targetRm == RoundingMode.UNNECESSARY || originalPrecision == 0) {
-                RoundingMode tmpRm =
-                    (targetRm == RoundingMode.UNNECESSARY) ? RoundingMode.DOWN : targetRm;
-                MathContext mcTmp = new MathContext(targetPrecision, tmpRm);
-                result = approx.scaleByPowerOfTen(-scaleAdjust/2).round(mcTmp);
-
-                // If result*result != this numerically, the square
-                // root isn't exact
-                if (this.subtract(result.square()).compareTo(ZERO) != 0) {
+                // After stripTrailingZeros, the representation is normalized as
+                //
+                // unscaledValue * 10^(-scale)
+                //
+                // where unscaledValue is an integer with the minimum
+                // precision for the cohort of the numerical value and the scale is even.
+                BigInteger[] sqrtRem = stripped.unscaledValue().sqrtAndRemainder();
+                result = new BigDecimal(sqrtRem[0], strippedScale >> 1);
+
+                // If result*result != this numerically or requires too high precision,
+                // the square root isn't exact
+                if (sqrtRem[1].signum != 0 || mc.precision != 0 && result.precision() > mc.precision)
                     throw new ArithmeticException("Computed square root not exact.");
-                }
+
+                // Test numerical properties at full precision before any
+                // scale adjustments.
+                assert squareRootResultAssertions(result, mc);
+                // Adjust to requested precision and preferred
+                // scale as appropriate.
+                result = result.adjustToPreferredScale(preferredScale, mc.precision);
```

Comment:
```suggestion
                return result.adjustToPreferredScale(preferredScale, mc.precision);
```
The redundant `else` below can then be removed, resulting in less "indented" code.
But the current choice is OK as well if you like it better.

Reply:
I've made the current choice for consistency with the old implementation, but indeed I think that using `return` is more handy.